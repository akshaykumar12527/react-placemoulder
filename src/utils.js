const REAPEAT_CHAR = ' .';

export const createObjectFromSchema = (schema) => {
  const outputObject = {};

  for (const key in schema) {
    const val = schema[key];
    if (val && typeof val === 'object') {
      // Recursively create object from schema object
      outputObject[key] = createObjectFromSchema(val);
    } else {
      outputObject[key] = REAPEAT_CHAR.repeat(+val);
    }
  }

  return outputObject;
};

export const mouldParagraph = (pTags) =>
  pTags.map(
    (pTag) =>
      (pTag.innerHTML = `${[...Array(3)].reduce(
        (resultStr, item) =>
          `${resultStr}<span class="stensil">${' .'.repeat(50)}</span>`,
        ''
      )}<span class="stensil half"/>`)
  );
