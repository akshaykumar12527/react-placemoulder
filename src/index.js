import React, { useRef, useEffect, useMemo } from 'react';

import { mouldParagraph, createObjectFromSchema } from './utils';

import './index.css';

export const Stencil = ({ children }) => {
  const currentRef = useRef(null);

  useEffect(() => {
    mouldParagraph([...currentRef.current.querySelectorAll('.stensil-para')]);
  }, []);

  return (
    <div className="enable-stensil" ref={currentRef}>
      {children}
    </div>
  );
};

export const StencilList = ({
  data,
  length,
  schema,
  Component,
}) => {
  // Create dummy props if config is provided
  const dummyProps = data || createObjectFromSchema(schema);

  // Memoize fakeArray based on length
  const fakeArray = useMemo(() => [...Array(length)], [length]);

  return fakeArray.map((_, index) => (
    <Stencil key={index}>
      <Component {...dummyProps} />
    </Stencil>
  ));
};
