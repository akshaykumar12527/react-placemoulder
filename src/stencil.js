import React, { useRef, useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';

import { mouldParagraph, createObjectFromSchema } from './utils';

import './stencil.css';

export const Stencil = ({ children }) => {
  const currentRef = useRef(null);

  useEffect(() => {
    mouldParagraph([...currentRef.current.querySelectorAll('.stensil-para')]);
  }, []);

  return (
    <div className="enable-stensil" ref={currentRef}>
      {children}
    </div>
  );
};

export const StencilList = ({
  data,
  length,
  schema,
  Component,
}) => {
  // Create dummy props if config is provided
  const dummyProps = data || createObjectFromSchema(schema);

  // Memoize fakeArray based on length
  const fakeArray = useMemo(() => [...Array(length)], [length]);

  return fakeArray.map((_, index) => (
    <Stencil key={index}>
      <Component {...dummyProps} />
    </Stencil>
  ));
};

Stencil.propTypes = {
  /** Children whose skeleton is required to be rendered */
  children: PropTypes.element.isRequired
}


StencilList.propTypes = {
  /** `data` is dummy or representational data which will be used to determine the dummy space occupied by the DOM element */
  data: PropTypes.object,

  /** Number of repetitive skeletons required to fill in the placeholder list */
  length: PropTypes.number.isRequired,

  /** If `data` is not provided, provide schema of the props required by `Component` prop. */
  schema: PropTypes.object,

  /** `Component` for which we want to generate skeleton on the fly  */
  Component: PropTypes.elementType.isRequired,
}