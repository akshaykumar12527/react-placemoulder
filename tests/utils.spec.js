import { createObjectFromSchema } from '../src/utils';

describe('Test utils', () => {
  describe('should test createObjectFromSchema', () => {
    it('should create object out schema with different supported data types', () => {
      const schema = {
        name: 10,
        address: {
          houseNo: 4,
          street: 80,
          locality: 20,
          country: 12,
          state: 12,
        },
        pin: null,
        resident: true,
        onSelect: (item) => {
          console.log('item');
        },
      };

      expect(createObjectFromSchema(schema)).toMatchSnapshot();
    });
  });
});
