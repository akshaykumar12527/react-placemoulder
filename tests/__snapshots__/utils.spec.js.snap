// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test utils should test createObjectFromSchema should create object out schema with different supported data types 1`] = `
Object {
  "address": Object {
    "country": " . . . . . . . . . . . .",
    "houseNo": " . . . .",
    "locality": " . . . . . . . . . . . . . . . . . . . .",
    "state": " . . . . . . . . . . . .",
    "street": " . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .",
  },
  "name": " . . . . . . . . . .",
  "onSelect": [Function],
  "resident": true,
}
`;
